chapter/methodology.md

```diff
@@ -0,0 +1,1023 @@
+# Design implementation of NLS
+
+This chapter guides through the various steps and components of the Nickel Language Server (NLS).
+Being written in the same language (Rust[@rust]) as the Nickel interpreter allows NLS to integrate existing components for language analysis.
+Aiming for an abstract interface, NLS defines its own data structure underpinning all higher level LSP interactions.
+[Section @sec:linearization] will introduce this `linearization` data structure and explain how NLS bridges the gap between the explicitly handled Nickel AST towards the abstract linearization.
+Finally, the implementation of current LSP features is discussed in [@sec:lsp-server].
+
+## Illustrative example
+
+The example [@lst:nickel-complete-example] shows an illustrative high level configuration of a server.
+Throughout this chapter, different sections about the NSL implementation will refer back to this example.
+
+```{.nickel #lst:nickel-complete-example caption="Nickel example with most features shown"}
+let Port | doc "A contract for a port number" =
+  contracts.from_predicate (fun value =>
+    builtins.is_num value &&
+    value % 1 == 0 &&
+    value >= 0 &&
+    value <= 65535) in
+
+let Container = {
+  image | Str,
+  ports | List Port,
+} in
+
+let NobernetesConfig = {
+  apiVersion | Str,
+  metadata.name | Str,
+  replicas | nums.PosNat
+           | doc "The number of replicas"
+           | default = 1,
+  containers | { _ : Container },
+
+} in
+
+let name_ = "myApp" in
+
+let metadata_ = {
+    name = name_,
+} in
+
+let webContainer = fun image => {
+  image = image,
+  ports = [ 80, 443 ],
+} in
+
+let image = "k8s.gcr.io/%{name_}" in
+
+{
+  apiVersion = "1.1.0",
+  metadata = metadata_,
+  replicas = 3,
+  containers = {
+    "main container" = webContainer image
+  }
+} | NobernetesConfig
+
+```
+
+
+## Linearization
+
+The focus of the NLS as presented in this work is to implement a working language server with a comprehensive feature set.
+To answer requests, NLS needs to store more information than what is originally present in a Nickel AST such as information about references and types.
+Apart from missing data, an AST is not optimized for quick random access of nodes based on their position, which is a crucial operation for a language server.
```

Comment:
As in "why is it 'a crucial operation'" i assume

Reply:
```suggestion
While these can be deduced from the AST lazily, it would require the repeated traversal of arbitrarily large tree with an associated cost to performance.
Therefore as hinted in [@sec:code-analysis], optimization is directed to efficient lookup from a pre-processed report.
Since most LSP commands refer to code positions, the intermediate structure must allow efficient lookup of analysis results based on positions.

```